src/k8s/pkg/k8sd/controllers/csrsigning/validate.go

```diff
@@ -9,19 +11,27 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-func validateCSR(obj *certv1.CertificateSigningRequest) error {
+func validateCSR(obj *certv1.CertificateSigningRequest, priv *rsa.PrivateKey) error {
 	csr, err := pkiutil.LoadCertificateRequest(string(obj.Spec.Request))
 	if err != nil {
 		return fmt.Errorf("failed to parse x509 certificate request: %w", err)
 	}
 
-	_ = csr
+	encryptedSignature := obj.Annotations["k8sd.io/signature"]
+	signature, err := rsa.DecryptPKCS1v15(nil, priv, []byte(encryptedSignature))
+	if err != nil {
+		return fmt.Errorf("failed to decrypt signature: %w", err)
+	}
 
-	// TODO(neoaggelos): validate requests have been encrypted using the k8sd public key:
-	// encryptedSignature = obj.Annotations["k8sd.io/signature"]
-	// signature = RSA_DECRYPT(k8sdPrivateKey, encryptedSignature)
-	// hash = SHA256(obj.Spec.Request)
-	// assert hash == signature
+	// calculate sha256 sum of CSR request
+	h := sha256.New()
+	if _, err := h.Write(obj.Spec.Request); err != nil {
+		return fmt.Errorf("failed to compute sha256: %w", err)
+	}
+
+	if !(string(h.Sum(nil)) == string(signature)) {
```

Comment:
this is written strangely, you can try `subtle.ConstantTimeCompare()` or just `A != B`